### DISCUSSION TITLE
How can I encode from video to VVC 10 bit and decode back from VVC to AV1?

### DISCUSSION THREAD

**Discussion by MartinEesmaa** [fraunhoferhhi/vvenc#102] [Q&A] [answered]
Hello guys.
I need help so, how can I encode from video to VVC 10 bit, example my input video codec is AOMEDIA 1 and I want encode into VVC 10 bit smaller bit rate without losing quality.
And after encode to VVC 10 bit, how can I decode back VVC to h.264, h.265 or av1?
Is there any best results to encode VVC 10 bit and decode back to AV1 to see media player VLC result.

- MartinHero

**Comment by adamjw24** [comment, marked as answer]
You cannot decode a video back to a different format, this is not a thing.

You can only decode the video from any format format back to raw video (raw YUV) and encode this raw YUV to another format. This is called transcoding. For VVenC, you'd need to actually create the intermediate YUV file or use piping, since it is not yet integrated into FFmpeg. For formats that have both encode and decode integration in FFmpeg, you don't need to create the intermedia files, since the decoding is done in FFmpeg memory, and the encoder uses it as a source.

If I understand correctly, you want to encode VVC, and transcode (decode to raw YUV and encode) it to some other format for viewing. You'd have to ensure that the transcoding does not introduce additional artifacts, so you'd need to use high-bitrate (the second encoding has to be visually lossless if you want to see the effects of VVC encoding). If you are using x265 for transcoding, use --cu-lossless parameter, and I'd propose a bitrate of 40mbps for HD and 100mbps for UHD. Don't know about AV1.

To keep the encodes and decodes as 10bit, make sure to use the yuv420_10 format in vvenc (vvdec will retain it), and after decoding your input format for the second encoding is yuv420p10le. Please check with your used codec how to ensure it encodes in 10bit, and visually lossless.

**Reply by MartinEesmaa** [reply]
Okay @adamjw24, so I did something else.
I tested my input file of .mp4, which is H.264 video codec / 4:2:0 8 bit.
I converted .mp4 into .yuv using MKVtoolnix GUI.
I encoded from yuv to .266 file using code `vvencapp -i a23.yuv -s 854x480 --format yuv420 -r 25 --threads 16 --preset slow --tier high -o whata23again.266`
VVC Log File:
[VVCLOGFILEmartinhero.txt](https://github.com/fraunhoferhhi/vvenc/files/7358464/VVCLOGFILEmartinhero.txt)

Did made output 10 bit after vvenc?
I tested using YUView and something went problem with vvdec, like this, same decoded file from 266 to YUV, same YUV and 266 file:
![YUViewerror](https://user-images.githubusercontent.com/88035011/137605759-18941166-6489-414d-8925-0052600b570d.png)


Reason: My all frames of 500 frames needs to be like same H.264 as VVC. It encoded 8 frames :/.

Anything you can give me solution?
- MartinHero13

**Reply by adamjw24** [reply]
Can you do a `dir` or `ls -l` in the working directory, to confirm the file sizes and be able to find the source of the problems?

Also, how does your .yuv generated in MKVtoolnix GUI look like in YUView? (it can show raw YUV)

**Reply by MartinEesmaa** [reply]
So, I checked and here's dir of folder VVC
![VVCFolderDIR](https://user-images.githubusercontent.com/88035011/137693132-39a89307-932e-4a98-9eac-dc637c655984.png)

While I trying to encode from yuv to .266 with VVENCAPP 1.2.0, then I got a error message:
![MicrosoftVISUALCERRORDEBUGVVC](https://user-images.githubusercontent.com/88035011/137693281-e1b1aeb7-f76b-4654-a48b-0365b6d1a1ca.png)

Oh, by the way, how did .yuv generated by MKVTOOLNIX GUI? Because, I converted from mp4 to yuv with no raw data.

- MartinHero13

**Reply by adamjw24** [reply]
First, please only encode in Release configuration.

Second, there is a problem with your YUV file (a23.yuv). It is 5170529 bytes large, which for 8 bit YUV with your resolution comes down to 8.409 frames (size / (W*H*1.5)). Please make sure your input data is correct, e.g. by previewing it in a YUV player.

**Reply by MartinEesmaa** [reply]
Okay, I kinda messed up. I downloaded releases tag from V 1.2.0, then I compiled with `cmake vvenc-1.2.0 (1)` then it compiled into whatever VSproject and other files.
I opened the project file: INSTALL.vcxproj
Then I run on Local Windows Debugger
After compiling, then I got a error, but it made success from vcxproj to 4 .exe files:
![Unabletostartaccessdenied1](https://user-images.githubusercontent.com/88035011/137698313-8dca47bb-39e3-4e4c-8f1e-c13a17881040.png)

Then I tried encode again from yuv to vvc, it did like same as error debug of Microsoft Visual C++ Runtime Library.
If you want my proof that I tried everything, here's my text file:
[ProofLOGFILEmartinheroVVC.txt](https://github.com/fraunhoferhhi/vvenc/files/7363117/ProofLOGFILEmartinheroVVC.txt)

- MartinHero13

**Comment by adamjw24** [comment]
Please refer to the README.txt or the whitepaper (see Wiki) for an explanation on how to build the project.

**Reply by MartinEesmaa** [reply]
I tried follow of the whitepaper and I did look same, but it's still stuck debug error.
I tried everything what you gave me solution, but it did not work :(, something is wrong with not well Microsoft Visual Studio.
I'm not good coding :(, sorry.
- MartinHero13

**Reply by MartinEesmaa** [reply]
UPDATE: Debug-static file report
Here's my debug static of still stuck debug error:
[ok2.zip](https://github.com/fraunhoferhhi/vvenc/files/7363369/ok2.zip)

If I'm sure that you can fix...
Inside .EXE, .PDB, .YUV and .266 file of ZIP.
- MartinHero13

**Reply by adamjw24** [reply]
On the top there in your Screenshot, select the "Release" in the drop-down menu which now says "Debug" (below the "Build" and "Debug" menu list options).

You can select which project to start in the top right "Solutions explorer" (select app->vvenc by right clicking and selecting "Set as exectuable"). In the properties you can select the parameters.

You can watch a tutorial on visual studio to figure it out.

Still the actual problem with your encode is your input YUV. Did you have a look at it?

**Reply by MartinEesmaa** [reply]
~~I just got a error called: ERROR: In function "vvenc::EncLib::encodePicture" and encoding failed code -1 after compiling build correctly Release x64 (only MKVtoolnixgui yuv converted file, not raw yuv file)~~
~~Here's pastebin report: https://pastebin.com/SBGkc1Rr~~

~~UPDATE FFMPEG LOG:~~
~~https://pastebin.com/FbD09Yfk~~

~~FFPLAY CORRUPTED YUV RAW FILE:~~

~~I opened the yuv raw file with VLC and it's corrupted, I just want to avoid corruption:~~
~~Here's ffmpeg that I used code: `ffmpeg -i INPUT8BIT.mp4 -c:v rawvideo -pixel_format yuv420p ok33.yuv`~~

UPDATE:
Okay good news, that I transcoded from video mp4 to yuv raw data using `ffmpeg -i INPUT8BIT.mp4 out.yuv`
I verified ffplay using: `ffplay -f rawvideo -pixel_format yuv420p -video_size 854x480 -i out.yuv` and it worked.
After verified of yuv, then I start encode using `vvencapp -i out.yuv -s 854x480 -r 30 --preset slow --threads max --tier high -o EXTREME.266`
After encoding of .266, then I start decode back to YUV, which makes me double size than original size: `vvdecapp -b EXTREME.266 -o backfromextreme.yuv`

Viewing backfromextreme.yuv (actually it was aj.yuv) in YUView:
![bigdoneforthat1](https://user-images.githubusercontent.com/88035011/137833446-b918cede-2658-4e14-a827-dfac89e76c14.png)

Thank you, @adamjw24! I will make sure to beginners encode to VVC file to let know. Before we gonna close this discussion, I am just trying to load .266 file in YUView, but it says corrupted, when I decoded from 266 file to YUV, it does not corrupt and it shows correctly.
![couldnotload123](https://user-images.githubusercontent.com/88035011/137838443-6510728d-8e4e-4823-bbda-797a309ef719.png)

Example of loaded .h266 file, that I wanna look this:
![simpleh266file](https://user-images.githubusercontent.com/88035011/137838737-63555419-d549-46a6-a7a0-376f04e8db44.png)

- MartinHero13

**Reply by adamjw24** [reply]
Hmm, thats strange. Can you decode the video with standalone vvdec? If so, it might be a problem with YUView. If not, we might have in vvdec and it would be great if you could provide some additional info.

**Reply by MartinEesmaa** [reply]
Okay, I solved it, because I ran on YUView develop version and I was successfully loaded my .266 file.
![itworked123](https://user-images.githubusercontent.com/88035011/138611004-7051a244-c3fd-48d4-9b8f-7e00beb7df7f.png)

Thank you, @adamjw24!
You can close this issue.
- MartinHero13
